File: scripts/performance/condense-candidates.py
#!/usr/bin/env python3
"""Compare complete native card journeys. No model calls; bytes are not tokens."""
import argparse
import hashlib
import json
import os
from pathlib import Path
import subprocess
import tempfile

ABOUT = 'project:card-replay'
ROOT, B, C = [f'{ABOUT}:{name}' for name in ('root', 'b', 'c')]
SOURCE = f'evidence:{ABOUT}:shared'


def encoded(value):
    return json.dumps(value, ensure_ascii=False, separators=(',', ':')).encode()


class Native:
    def __init__(self, binary, directory):
        env = {k: v for k, v in os.environ.items() if not k.startswith('KMP_')}
        env.update(KMP_MCP_BACKEND='embedded', KMP_MCP_DATA_DIR=str(directory / 'store'),
                   KMP_VIEWER_ADDR='127.0.0.1:0')
        self.process = subprocess.Popen([str(binary)], cwd=directory, env=env,
            stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL)
        self.rows = []

    def rpc(self, method, params, phase):
        request = encoded(dict(jsonrpc='2.0', id=len(self.rows)+1, method=method, params=params))
        self.process.stdin.write(request+b'\n')
        self.process.stdin.flush()
        raw = self.process.stdout.readline().rstrip(b'\r\n')
        if not raw:
            raise RuntimeError(f'No response from native binary during {phase}')
        response = json.loads(raw)
        result = response.get('result', {})
        if 'error' in response or result.get('isError'):
            raise RuntimeError(response)
        field = result.get('structuredContent', {}).get('proof', {}).get('condense_candidates')
        self.rows.append(dict(phase=phase, request_bytes=len(request), response_bytes=len(raw),
                              candidate_field_bytes=len(encoded(field)) if field is not None else 0))
        return result

    def call(self, name, arguments, phase):
        return self.rpc('tools/call', dict(name=name, arguments=arguments), phase)['structuredContent']

    def pages(self, arguments, phase):
        results = []
        while True:
            result = self.call('kmp_trace', arguments, phase)
            results.append(result)
            if not result.get('page', {}).get('has_more'):
                return results
            assert len(results) < 100
            arguments = result['next_actions'][0]['arguments']

    def close(self):
        self.process.stdin.close()
        try:
            self.process.wait(timeout=10)
        except subprocess.TimeoutExpired:
            self.process.terminate()
            self.process.wait(timeout=10)


def seed(size):
    entries = [dict(id=ref, kind='observation', text='entry '+('e'*size),
        coordinates=[dict(dimension='task', scope_id='read', occurred_at='2026-07-22T10:00:00Z')])
        for ref in [ROOT, B, C]]
    relations = [dict(**{'from': ROOT, 'to': ref}, rel='depends_on',
        **{'class': 'causal'}, confidence='high', why='The source declares the dependency.',
        evidence='Source register.') for ref in [B, C]]
    return dict(about=ABOUT, idempotency_key='replay', memory=dict(
        dimensions=[dict(id='read', kind='task')], entries=entries, relations=relations,
        evidence=[dict(id=SOURCE, supports=[ROOT,B,C], text='shared '+('s'*(size*4)), source='signed register')]))


def run(binary, scratch, size, mode):
    with tempfile.TemporaryDirectory(prefix='card-replay-', dir=scratch) as directory:
        client = Native(binary, Path(directory))
        try:
            client.rpc('initialize', dict(protocolVersion='2024-11-05', capabilities={},
                clientInfo=dict(name='condense-candidates-replay',version='1')), 'initialize')
            client.rpc('tools/list', {}, 'tools_list')
            client.call('kmp_ingest', seed(size), 'seed')
            query = dict(about=ABOUT, **{'from':ROOT, 'to':[B,C]},
                search=dict(proof=True, compact=dict(language='en')),
                page=dict(entries=2), budget=dict(max_bytes=200000))
            discovery = client.pages(query, 'discovery')
            objects = [o for page in discovery for o in page.get('objects', [])]
            if mode == 'candidate':
                choices = discovery[0]['proof']['condense_candidates']['items']
                if choices:
                    target = choices[0]
                    assert target['ref'] == SOURCE
                    source, expect = target['source'], target['expect']
                    record_bytes = target['record_bytes']
                else:
                    # The small-body negative control deliberately exercises an
                    # unoffered target; this does not attribute benefit to the list.
                    target = next(o for o in objects if o['ref'] == SOURCE)
                    source = {k:target['descriptor'][k] for k in ['revision','record_digest']}
                    expect, record_bytes = dict(absent=True), target['descriptor']['record_bytes']
            else:
                # A scripted full scan supplies the same target, without claiming
                # to model a reader's target choice or identity-copy error rate.
                target = next(o for o in objects if o['ref'] == SOURCE)
                source = {k:target['descriptor'][k] for k in ['revision','record_digest']}
                expect, record_bytes = dict(absent=True), target['descriptor']['record_bytes']
            expand = json.loads(json.dumps(query))
            expand['search'].update(proof_refs=[SOURCE],
                expect_selection=discovery[0]['proof']['manifest_id'], max_body_record_bytes=record_bytes)
            expansion = client.pages(expand, 'expansion')
            body = next(o['text'] for p in expansion for o in p.get('objects', []) if o['ref']==SOURCE)
            assert body == seed(size)['memory']['evidence'][0]['text']
            client.call('kmp_condense', dict(about=ABOUT, ref=SOURCE, scope='node_body', language='en',
                card='The signed register supports the three entries.', source=source, expect=expect,
                actor='replay-reader'), 'condense')
            fresh = client.pages(query, 'fresh_compact')
            assert fresh[0]['proof']['compact']['valid']==1
            journey = [r for r in client.rows if r['phase'] not in ['initialize','tools_list','seed']]
            return dict(mode=mode, entry_filler_bytes=size, binary_sha256=hashlib.sha256(binary.read_bytes()).hexdigest(),
                calls=len(journey), request_bytes=sum(r['request_bytes'] for r in journey),
                response_bytes=sum(r['response_bytes'] for r in journey),
                total_bytes=sum(r['request_bytes']+r['response_bytes'] for r in journey),
                candidate_field_bytes=sum(r['candidate_field_bytes'] for r in journey),
                source_sha256=hashlib.sha256(body.encode()).hexdigest(), phases=client.rows)
        finally:
            client.close()


def main():
    parser=argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--baseline',type=Path,required=True)
    parser.add_argument('--candidate',type=Path,required=True)
    parser.add_argument('--scratch',type=Path,required=True)
    parser.add_argument('--output',type=Path,required=True)
    args=parser.parse_args()
    args.scratch.mkdir(parents=True,exist_ok=True)
    results=[]
    for size in [4096, 120]:
        for mode in ['baseline','candidate']:
            results.append(run(getattr(args,mode).resolve(),args.scratch.resolve(),size,mode))
    for baseline,candidate in zip(results[::2],results[1::2]):
        assert baseline['source_sha256']==candidate['source_sha256']
    args.output.write_text(json.dumps(dict(contract='condense-candidate-replay-v1',
        scope='Complete JSON-RPC requests/responses, without line delimiters. No model, token or latency claim.',
        results=results),indent=2)+'\n')
    print(json.dumps([{k:v for k,v in r.items() if k!='phases'} for r in results],indent=2))


if __name__=='__main__':
    main()

File: crates/kmp-mcp/tests/condense_candidates.rs
//! Card targeting through the public MCP surface and a real SQLite snapshot.
use kmp_mcp::{EmbeddedKernelMcpBackend, KernelMcpServer};
use serde_json::{Value, json};

const ABOUT: &str = "project:condense-candidates";
const ROOT: &str = "project:condense-candidates:root";
const B: &str = "project:condense-candidates:b";
const C: &str = "project:condense-candidates:c";
const SOURCE: &str = "evidence:project:condense-candidates:shared";
const EARLY: &str = "2026-07-22T10:00:00Z";

async fn call(server: &KernelMcpServer, name: &str, arguments: Value) -> Value {
    let raw = server
        .handle_json_line(
            &json!({
                "jsonrpc":"2.0", "id":1, "method":"tools/call",
                "params":{"name":name,"arguments":arguments}
            })
            .to_string(),
        )
        .await
        .expect("valid candidate fixture");
    let response: Value = serde_json::from_str(&raw).expect("valid candidate fixture");
    response["result"].clone()
}

fn content(result: &Value) -> &Value {
    assert_ne!(result["isError"], true, "{result}");
    &result["structuredContent"]
}

fn packet(generation: usize) -> Value {
    let entries = [(ROOT, 5000), (B, 2048), (C, 500)].map(|(id, size)| json!({
        "id":id,"kind":"observation","text":"x".repeat(size),
        "coordinates":[{"dimension":"task","scope_id":"read","occurred_at":EARLY,"observed_at":EARLY}]
    }));
    let relations = [B, C].map(|id| {
        json!({
            "from":ROOT,"to":id,"rel":"depends_on","class":"causal","confidence":"high",
            "why":"Both branches require the signed source.","evidence":"Signed source register.",
            "clocks":{"occurred_at":EARLY,"observed_at":EARLY}
        })
    });
    json!({
        "about":ABOUT,"idempotency_key":format!("seed-{generation}"),
        "memory":{
            "dimensions":[{"id":"read","kind":"task"}],"entries":entries,"relations":relations,
            "evidence":[
                {"id":SOURCE,"supports":[B,C],"text":"s".repeat(8192+generation),"source":"signed register","time":EARLY,
                    "support_clocks":{"observed_at":EARLY}},
                {"id":"evidence:project:condense-candidates:root","supports":[ROOT],"text":"Root register.","source":"root register","time":EARLY,
                    "support_clocks":{"observed_at":EARLY}}
            ]
        }
    })
}

async fn seed(server: &KernelMcpServer, generation: usize) {
    content(&call(server, "kmp_ingest", packet(generation)).await);
}

async fn server() -> (tempfile::TempDir, KernelMcpServer) {
    let dir = tempfile::tempdir().expect("valid candidate fixture");
    let server = KernelMcpServer::with_embedded_backend(
        EmbeddedKernelMcpBackend::open(dir.path()).expect("valid candidate fixture"),
    );
    seed(&server, 0).await;
    (dir, server)
}

fn request() -> Value {
    json!({"about":ABOUT,"from":ROOT,"to":[B,C],
        "search":{"proof":true,"compact":{"language":"es"}},
        "budget":{"max_bytes":200_000}})
}

fn write_from(candidate: &Value) -> Value {
    json!({"about":ABOUT,"ref":candidate["ref"],"scope":"node_body","language":"es",
        "source":candidate["source"],"expect":candidate["expect"],
        "actor":"candidate-test", "card":"El registro firmado respalda ambas ramas."})
}

#[tokio::test]
async fn every_page_ranks_the_shared_source_and_its_identity_can_be_used_directly() {
    let (_dir, server) = server().await;
    let whole = call(&server, "kmp_trace", request()).await;
    let expected = content(&whole)["proof"]["condense_candidates"].clone();
    assert_eq!(expected["items"][0]["ref"], SOURCE);
    assert_eq!(expected["items"][0]["shared_by"], 2);
    assert_eq!(expected["items"][0]["body_bytes"], 8192);
    assert_eq!(expected["below_floor"], 2);
    assert_eq!(content(&whole)["proof"]["body_bytes"], 0);
    assert_eq!(content(&whole)["proof"]["complete_groups"], json!([]));

    let mut paged = request();
    paged["page"] = json!({"entries":1});
    let mut pages = 0;
    loop {
        let result = call(&server, "kmp_trace", paged).await;
        let value = content(&result);
        assert_eq!(value["proof"]["condense_candidates"], expected);
        pages += 1;
        assert!(pages < 50);
        if !value["page"]["has_more"]
            .as_bool()
            .expect("valid candidate fixture")
        {
            break;
        }
        paged = value["next_actions"][0]["arguments"].clone();
    }
    assert!(pages > 2);

    // Read the canonical body before authoring. No descriptor transcribing.
    let mut expansion = request();
    expansion["search"]["proof_refs"] = json!([SOURCE]);
    expansion["search"]["max_body_record_bytes"] = expected["items"][0]["record_bytes"].clone();
    expansion["search"]["expect_selection"] = content(&whole)["proof"]["manifest_id"].clone();
    let expanded = call(&server, "kmp_trace", expansion).await;
    let source = content(&expanded)["objects"]
        .as_array()
        .expect("valid candidate fixture")
        .iter()
        .find(|o| o["ref"] == SOURCE)
        .expect("valid candidate fixture");
    assert_eq!(
        source["text"]
            .as_str()
            .expect("valid candidate fixture")
            .len(),
        8192
    );
    let candidate = &expected["items"][0];
    assert_eq!(
        candidate["source"]["revision"],
        source["descriptor"]["revision"]
    );
    assert_eq!(
        candidate["source"]["record_digest"],
        source["descriptor"]["record_digest"]
    );
    let written = call(&server, "kmp_condense", write_from(candidate)).await;
    content(&written);
    let compact = call(&server, "kmp_trace", request()).await;
    let candidates = &content(&compact)["proof"]["condense_candidates"];
    assert_eq!(candidates["valid"], 1);
    assert!(
        candidates["items"]
            .as_array()
            .expect("valid candidate fixture")
            .iter()
            .all(|c| c["ref"] != SOURCE)
    );

    let mut other_language = request();
    other_language["search"]["compact"]["language"] = json!("en");
    let english = call(&server, "kmp_trace", other_language).await;
    let first = &content(&english)["proof"]["condense_candidates"]["items"][0];
    assert_eq!(first["ref"], SOURCE);
    assert_eq!(first["expect"], json!({"absent":true}));

    // The same stored card is unavailable on a historical read.
    let mut historical = request();
    historical["axis"] = json!("occurred");
    historical["as_of"] = json!({"time":"2026-08-01T00:00:00Z"});
    let past = call(&server, "kmp_trace", historical).await;
    let candidates = &content(&past)["proof"]["condense_candidates"];
    assert_eq!(candidates["after_cut"], 1, "{past}");
    assert!(
        candidates["items"]
            .as_array()
            .expect("valid candidate fixture")
            .iter()
            .all(|c| c["ref"] != SOURCE)
    );
}

#[tokio::test]
async fn moved_source_is_refused_and_a_stale_card_carries_the_exact_cas_revision() {
    let (_dir, server) = server().await;
    let before = call(&server, "kmp_trace", request()).await;
    let old = content(&before)["proof"]["condense_candidates"]["items"][0].clone();
    content(&call(&server, "kmp_condense", write_from(&old)).await);
    seed(&server, 1).await;
    let refused = call(&server, "kmp_condense", write_from(&old)).await;
    assert_eq!(refused["isError"], true);
    assert_eq!(refused["structuredContent"]["error"]["code"], "conflict");
    assert!(
        refused.to_string().contains("body of this entry moved"),
        "{refused}"
    );
    let now = call(&server, "kmp_trace", request()).await;
    let fresh = &content(&now)["proof"]["condense_candidates"]["items"][0];
    assert_eq!(fresh["ref"], SOURCE);
    assert_eq!(fresh["card_status"], "stale");
    assert_eq!(fresh["expect"], json!({"card_revision":1}));
    assert_ne!(fresh["source"], old["source"]);
    content(&call(&server, "kmp_condense", write_from(fresh)).await);
    let concurrent = call(&server, "kmp_condense", write_from(fresh)).await;
    assert_eq!(concurrent["isError"], true);
    assert_eq!(concurrent["structuredContent"]["error"]["code"], "conflict");
    assert!(concurrent.to_string().contains("stored card revision 2"));
}

#[tokio::test]
async fn noncompact_reads_and_refused_expansions_do_not_offer_candidates() {
    let (_dir, server) = server().await;
    let mut plain = request();
    plain["search"]
        .as_object_mut()
        .expect("valid candidate fixture")
        .remove("compact");
    let read = call(&server, "kmp_trace", plain).await;
    assert!(content(&read)["proof"].get("condense_candidates").is_none());
    let mut unknown = request();
    unknown["search"]["proof_refs"] = json!(["foreign:ref"]);
    let compact = call(&server, "kmp_trace", request()).await;
    unknown["search"]["expect_selection"] = content(&compact)["proof"]["manifest_id"].clone();
    unknown["search"]["max_body_record_bytes"] = json!(100_000);
    let refused = call(&server, "kmp_trace", unknown).await;
    assert!(
        content(&refused)["proof"]
            .get("condense_candidates")
            .is_none()
    );
}

#[tokio::test]
async fn seek_counts_complete_structural_groups_before_compact_body_delivery() {
    let (_dir, server) = server().await;
    let mut seek = request();
    seek.as_object_mut()
        .expect("valid candidate fixture")
        .remove("to");
    seek["search"]["seek"] = json!(["depends_on"]);
    let result = call(&server, "kmp_trace", seek).await;
    let value = content(&result);
    assert_eq!(
        value["groups"]
            .as_array()
            .expect("valid candidate fixture")
            .len(),
        2,
        "{value}"
    );
    let candidates = &value["proof"]["condense_candidates"]["items"];
    assert_eq!(candidates[0]["ref"], SOURCE);
    assert_eq!(candidates[0]["shared_by"], 2);
    assert_eq!(value["proof"]["complete_groups"], json!([]));
}

#[tokio::test]
async fn seek_does_not_count_groups_with_missing_witnesses_as_shared_proof() {
    let (_dir, server) = server().await;
    let mut seek = request();
    seek.as_object_mut()
        .expect("valid candidate fixture")
        .remove("to");
    seek["search"]["seek"] = json!(["depends_on"]);
    seek["search"]["same_labels"] = json!(["event"]);
    let result = call(&server, "kmp_trace", seek).await;
    let value = content(&result);
    assert_eq!(value["seek"]["status"], "review_required", "{value}");
    assert_eq!(value["groups"].as_array().expect("groups").len(), 2);
    let candidates = value["proof"]["condense_candidates"]["items"]
        .as_array()
        .expect("candidates");
    assert!(!candidates.is_empty());
    assert!(
        candidates.iter().all(|c| c["shared_by"] == 0),
        "{candidates:?}"
    );
}

#[tokio::test]
async fn historical_seek_does_not_count_groups_with_unknown_relation_clocks() {
    let dir = tempfile::tempdir().expect("temporary store");
    let server = KernelMcpServer::with_embedded_backend(
        EmbeddedKernelMcpBackend::open(dir.path()).expect("SQLite backend"),
    );
    let mut seed = packet(0);
    for edge in seed["memory"]["relations"]
        .as_array_mut()
        .expect("relations")
    {
        edge.as_object_mut().expect("relation").remove("clocks");
    }
    content(&call(&server, "kmp_ingest", seed).await);
    let mut seek = request();
    seek.as_object_mut().expect("request").remove("to");
    seek["search"]["seek"] = json!(["depends_on"]);
    seek["axis"] = json!("occurred");
    seek["as_of"] = json!({"time":"2026-08-01T00:00:00Z"});
    let result = call(&server, "kmp_trace", seek).await;
    let value = content(&result);
    assert_eq!(value["seek"]["status"], "review_required", "{value}");
    let groups = value["groups"].as_array().expect("groups");
    assert_eq!(groups.len(), 2);
    assert!(groups.iter().all(|g| g["clock_unknown"] == true));
    let candidates = value["proof"]["condense_candidates"]["items"]
        .as_array()
        .expect("candidates");
    assert!(!candidates.is_empty());
    assert!(
        candidates.iter().all(|c| c["shared_by"] == 0),
        "{candidates:?}"
    );
}

#[tokio::test]
async fn material_selection_counts_only_returned_routes() {
    let (_dir, server) = server().await;
    let mut selected = request();
    selected["search"]["select"] = json!({"max_material_nodes":2,"max_paths":1});
    let result = call(&server, "kmp_trace", selected).await;
    let value = content(&result);
    assert_eq!(value["routes"].as_array().expect("routes").len(), 1);
    assert_eq!(value["search"]["material"]["candidate_count"], 2);
    let candidates = value["proof"]["condense_candidates"]["items"]
        .as_array()
        .expect("candidates");
    assert!(!candidates.is_empty());
    assert!(
        candidates.iter().all(|c| c["shared_by"] == 1),
        "{candidates:?}"
    );
}
